smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/action/SmallFileCompactAction.java

```diff
@@ -0,0 +1,252 @@
+
+  @Override
+  public void init(Map<String, String> args) {
+    super.init(args);
+    this.conf = getContext().getConf();
+    this.xAttrName = conf.get(
+        SmartConfKeys.SMART_FILE_STATE_XATTR_NAME_KEY,
+        SmartConfKeys.SMART_FILE_STATE_XATTR_NAME_DEFAULT);
+    this.smallFiles = args.get(FILE_PATH);
+    this.containerFile = args.get(CONTAINER_FILE);
+    this.containerFilePermission = args.get(CONTAINER_FILE_PERMISSION);
+  }
+
+  @Override
+  protected void execute() throws Exception {
+    // Set hdfs client by DFSClient rather than SmartDFSClient
+    this.setDfsClient(HadoopUtil.getDFSClient(
+        HadoopUtil.getNameNodeUri(conf), conf));
+
+    // Get small file list
+    if (smallFiles == null || smallFiles.isEmpty()) {
+      throw new IllegalArgumentException(
+          String.format("Invalid small files: %s.", smallFiles));
+    }
+    ArrayList<String> smallFileList = new Gson().fromJson(
+        smallFiles, new TypeToken<ArrayList<String>>() {
+        }.getType());
+
+    // Get container file path
+    if (containerFile == null || containerFile.isEmpty()) {
+      throw new IllegalArgumentException(
+          String.format("Invalid container file: %s.", containerFile));
+    }
+
+    // Get container file permission
+    SmartFilePermission filePermission = null;
+    if (containerFilePermission != null && !containerFilePermission.isEmpty()) {
+      filePermission = new Gson().fromJson(
+          containerFilePermission, new TypeToken<SmartFilePermission>() {
+          }.getType());
+    }
+
+    // Get initial offset and output stream
+    // Create container file and set permission if not exists
+    long offset;
+    OutputStream out;
+    if (exists(containerFile)) {
+      offset = getFileLength(containerFile);
+      out = getAppendOutputStream(containerFile);
+    } else {
+      out = dfsClient.create(containerFile, true);
+      if (filePermission != null) {
+        dfsClient.setOwner(
+            containerFile, filePermission.getOwner(), filePermission.getGroup());
+        dfsClient.setPermission(
+            containerFile, new FsPermission(filePermission.getPermission()));
+      }
+      offset = 0L;
+    }
+
+    appendLog(String.format("Action starts at %s : compact small files to %s.",
+        Utils.getFormatedCurrentTime(), containerFile));
+    for (String smallFile : smallFileList) {
+      if ((smallFile != null) && !smallFile.isEmpty()
+          && exists(smallFile) && getFileLength(smallFile) > 0) {
+        try (InputStream in = getInputStream(smallFile)) {
+          // Copy bytes of small file to container file
+          IOUtils.copyBytes(in, out, 4096);
+
+          // Truncate small file and add file container info to XAttr
+          long fileLen = getFileLength(smallFile);
+          FileState compactFileState = new CompactFileState(
+              smallFile, new FileContainerInfo(containerFile, offset, fileLen));
+          truncate0(smallFile, compactFileState);
```

Comment:
What about the case: 1st file success, but the 2th file fails, then the 1st file has an x-attribute which may be invalid.